_posts/2022-05-09-project-loom-for-distributed-systems.md

```diff
@@ -0,0 +1,397 @@
+
+Java's Project Loom considerably shakes up this tradeoff.
+
+![spectrum](/assets/images/spectrum.jpg){: width="100%" }
+
+# Project Loom
+
+[Project Loom](https://wiki.openjdk.java.net/display/loom/Main) provides 'virtual' threads as a first class concept within Java. There is plenty of good information in the 2020 blog post '[State of Loom](http://cr.openjdk.java.net/~rpressler/loom/loom/sol1_part1.html)' although details have changed in the last two years.
+
+I will give a simplified description of what I find exciting about this. In traditional Java we can launch a Thread. If it needs to pause for some reason, the thread will be paused, and will resume when it is able to. Java does not make it easy to control the threads (pause at a critical section, pick who acquired the lock, etc), and so influencing the interleaving of execution is very difficult except for in very isolated cases.
+
+With Loom's virtual threads, when a thread starts, a `Runnable` is submitted to an `Executor`. When that task is run by the executor, if the thread needs to block, the submitted runnable will exit, instead of pausing. When the thread can be unblocked, a new runnable is submitted to the same executor to pick up where the previous `Runnable` left off. Here, interleaving is much, much easier, since we are passed each piece of runnable work as it becomes runnable. Combined with the `Thread.yield()` primitive, we can also influence the points at which code becomes deschedulable.
+
+```java
+Queue<Runnable> executor = new ArrayDeque<>();
+Lock lock = new ReentrantLock();
+lock.lock();
+newVirtualThread(executor::add, lock::lock);
+assertThat(executor).hasSize(1)
+    .as("runnable for vthread has been submitted");
+executor.poll().run();
+assertThat(executor).hasSize(0)
+    .as("vthread has blocked, no longer runnable");
+lock.unlock();
+assertThat(executor).hasSize(1)
+    .as("due to unlock, the vthread is now schedulable again");
+executor.poll().run();
+assertThat(lock.tryLock()).isFalse()
+    .as("the virtual thread now holds the lock");
+```
+
+By utilizing this API, we can exert fine grained deterministic control over execution within Java. Let's give a micro level example. Suppose we're trying to test the correctness of a buggy version of Guava's `Suppliers.memoize` function.
+
+```java
+class BrokenMemoizingSupplier<T> implements Supplier<T> {
+  private final Lock lock;
+  private final Supplier<T> delegate;
+  private volatile boolean initialized = false;
+  private volatile T value;
+
+  BrokenMemoizingSupplier(Supplier<T> delegate) {
+    this(new ReentrantLock(), delegate);
+  }
+
+  @VisibleForTesting
+  BrokenMemoizingSupplier(Lock lock, Supplier<T> delegate) {
+    this.lock = lock;
+    this.delegate = delegate;
+  }
+
+  @Override
+  public T get() {
+    if (!initialized) {
+      lock.lock();
+      try {
+        // This code is broken because initialized may have changed between its check before the call to 'lock' and now.
```

Comment:
worth adding one more sentence here to explain what exactly is broken: delegate.get() is called more than once. (yes, the test below clarifies this, but would still add the comment here)